zulipterminal/model.py

```diff
@@ -584,13 +588,45 @@ def get_latest_message_in_topic(
         display_error_if_present(response, self.controller)
         return response
 
+    def can_user_edit_topic(self) -> Optional[str]:
+        user_info = self.get_user_info(self.user_id)
+        if user_info is not None:
+            role = user_info["role"]
+            if role is not None:
+                if self.initial_data.get("realm_edit_topic_policy") == 2:
+                    # For admins and owners only (not using user role)
+                    if role <= 200:
+                        return None
+                    else:
+                        return TOPIC_EDIT_POLICY[2]
+                elif self.initial_data.get("realm_edit_topic_policy") == 4:
+                    # Check for moderators only stream
+                    if role <= 300:
+                        return None
+                    else:
+                        return TOPIC_EDIT_POLICY[4]
+                elif self.initial_data.get("realm_edit_topic_policy") == 3:
+                    # Check for 'members' in the stream
+                    if role <= 400:
+                        return None
+                    else:
+                        return TOPIC_EDIT_POLICY[3]
+                else:
+                    # Anyone can post
+                    return None
+        return "User not found"
+
     def toggle_topic_resolved_status(self, stream_id: int, topic_name: str) -> None:
+        user_topic_permission = self.can_user_edit_topic()
         if not self.initial_data.get("realm_allow_message_editing", None):
             self.controller.report_error(" Editing messages is disabled")
             return
         elif self.initial_data.get("realm_allow_community_topic_editing", None):
             self.controller.report_error(" Editing topic is disabled")
             return
+        elif user_topic_permission is not None:
+            self.controller.report_error(user_topic_permission)
+            return
         else:
             response = self.get_latest_message_in_topic(stream_id, topic_name)
```

Comment:
You introduce `can_user_edit_topic` in this commit, but it's only handling a feature which I think is new (check the zulip API changelog). That it's looking to extend the base functionality to handle a new feature is great! The reason I've highlighted this section is since all the error outputs due to permissions in toggle_topic_resolved_status are also relevant to this, so I'd suggest adding can_user_edit_topic in the first related commit, then extending it here when you add the functionality only available in updated servers.